### PR TITLE
Support multi-group mRNA and protein enrichments

### PR DESCRIPTION
**Pull request by kalletlak**
# What? Why?
frontend changes for https://github.com/cBioPortal/cbioportal/issues/6105

### CODE REVIEW

File: src/pages/groupComparison/GroupComparisonStore.ts [hunk 1]
```diff
@@ -664,18 +654,29 @@ export default class GroupComparisonStore {
...
+                                molecularProfileId: molecularProfileId
+                            }
+                        });
+                    });
+                    return {
+                        name: group.nameWithOrdinal,
+                        molecularProfileCaseIdentifiers
+                    }
+                });
                 return internalClient.fetchExpressionEnrichmentsUsingPOST({
```
Comment (adamabeshouse) [inline src/pages/groupComparison/GroupComparisonStore.ts]: you should break here and return empty if groups.length < 2

Comment (adamabeshouse) [reply src/pages/groupComparison/GroupComparisonStore.ts]: same for proteinEnrichmentData

File: src/pages/groupComparison/GroupComparisonStore.ts [hunk 2]
```diff
@@ -766,7 +779,6 @@ export default class GroupComparisonStore {
     @computed get mRNATabGrey() {
         // grey out if
         return (this.activeStudyIds.isComplete && this.activeStudyIds.result.length > 1) // more than one active study
-            || (this.activeGroups.isComplete && this.activeGroups.result.length !== 2) // not two active groups
```
Comment (adamabeshouse) [inline src/pages/groupComparison/GroupComparisonStore.ts]: less than two active groups, right?

File: src/pages/groupComparison/GroupComparisonStore.ts [hunk 3]
```diff
@@ -782,7 +794,6 @@ export default class GroupComparisonStore {
     @computed get proteinTabGrey() {
         // grey out if
         return (this.activeStudyIds.isComplete && this.activeStudyIds.result.length > 1) // more than one active study
-            || (this.activeGroups.isComplete && this.activeGroups.result.length !== 2) // not two active groups
```
Comment (adamabeshouse) [inline src/pages/groupComparison/GroupComparisonStore.ts]: less than two active groups

File: src/pages/resultsView/enrichments/MRNAEnrichmentsTab.tsx [hunk 4]
```diff
@@ -35,6 +36,21 @@ export default class MRNAEnrichmentsTab extends React.Component<IMRNAEnrichments
                         selectedValue={this.props.store.selectedEnrichmentMRNAProfile.molecularProfileId}
                         molecularProfileIdToProfiledSampleCount={this.props.store.molecularProfileIdToProfiledSampleCount}/>
                     <ExpressionEnrichmentContainer data={this.props.store.mRNAEnrichmentData.result!}
+                    groups={[
+                        {
+                            name: "Altered group",
+                            description: "samples that have alterations in the query gene(s).",
+                            nameOfEnrichmentDirection: "Over-expressed",
+                            count: this.props.store.alteredSampleKeys.result!.length,
+                            color: this.uniqueColorGetter()
```
Comment (adamabeshouse) [inline src/pages/resultsView/enrichments/MRNAEnrichmentsTab.tsx]: get these colors once up top as private members because otherwise every time the component reloads they'll be different colors. same for protein

File: src/pages/groupComparison/GroupComparisonStore.ts (hunk 1 above)
Comment (kalletlak) [reply src/pages/groupComparison/GroupComparisonStore.ts]: it won't be invoked if groups.length < 2

https://github.com/cBioPortal/cbioportal-frontend/blob/4049894b6b08d52e6d1f04c7794360d9a110f8c7/src/pages/groupComparison/GroupComparisonUtils.tsx#L321

File: src/pages/resultsView/enrichments/ProteinEnrichmentsTab.tsx [hunk 5]
```diff
@@ -24,35 +24,41 @@ export default class ProteinEnrichmentsTab extends React.Component<IProteinEnric
...
 
+    private readonly enrichmentAnalysisGroups = remoteData({
+        await: () => [this.props.store.unalteredSampleKeys],
+        invoke: () => {
+            const groups = [{
+                name: "Altered group",
+                description: "samples that have alterations in the query gene(s).",
+                nameOfEnrichmentDirection: "Over-expressed",
+                count: this.props.store.alteredSampleKeys.result!.length,
+                color: this.uniqueColorGetter()
```
Comment (adamabeshouse) [inline src/pages/resultsView/enrichments/ProteinEnrichmentsTab.tsx]: I think the same problem will happen. You should recreate uniqueColorGetter on every invoke to avoid it

File: src/pages/groupComparison/GroupComparisonStore.ts [hunk 6]
```diff
@@ -684,18 +680,25 @@ export default class GroupComparisonStore {
...
-        await:()=>[this.enrichmentsGroup1, this.enrichmentsGroup2,this.proteinEnrichmentProfile],
+        await:()=>[this.proteinEnrichmentProfile],
         getSelectedProfile:()=>this.proteinEnrichmentProfile.result,// returns an empty array if the selected study doesn't have any mRNA profiles
         fetchData:()=>{
             // assumes single study for now
-            if (this.enrichmentsGroup1.result && this.enrichmentsGroup2.result && this.proteinEnrichmentProfile.result) {
+            if (this.proteinEnrichmentProfile.result) {
+                const molecularProfileId = this.proteinEnrichmentProfile.result!.molecularProfileId;
+                const groups: MolecularProfileCasesGroupFilter[] = _.map(this._activeGroupsOverlapRemoved.result, group => {
+                    const molecularProfileCaseIdentifiers = _.flatMap(group.studies, study => {
```
Comment (alisman) [inline src/pages/groupComparison/GroupComparisonStore.ts]: should this we awaited (_activeGroupsOverlapRemoved)?

File: src/pages/groupComparison/GroupComparisonStore.ts [hunk 7]
```diff
@@ -764,9 +767,8 @@ export default class GroupComparisonStore {
     }
 
     @computed get mRNATabGrey() {
-        // grey out if
-        return (this.activeStudyIds.isComplete && this.activeStudyIds.result.length > 1) // more than one active study
-            || (this.activeGroups.isComplete && this.activeGroups.result.length !== 2) // not two active groups
+        return (this.activeGroups.isComplete && this.activeGroups.result.length < 2) //less than two active groups
```
Comment (alisman) [inline src/pages/groupComparison/GroupComparisonStore.ts]: lets change the name of this to mRNATabDisabled or something.  so it's presentation agnostic

File: src/pages/groupComparison/GroupComparisonStore.ts [hunk 8]
```diff
@@ -780,9 +782,8 @@ export default class GroupComparisonStore {
     }
 
     @computed get proteinTabGrey() {
-        // grey out if
-        return (this.activeStudyIds.isComplete && this.activeStudyIds.result.length > 1) // more than one active study
-            || (this.activeGroups.isComplete && this.activeGroups.result.length !== 2) // not two active groups
+        return (this.activeGroups.isComplete && this.activeGroups.result.length < 2) //less than two active groups
```
Comment (alisman) [inline src/pages/groupComparison/GroupComparisonStore.ts]: same here

File: src/pages/groupComparison/ProteinEnrichments.tsx [hunk 9]
```diff
@@ -21,31 +23,39 @@ export default class ProteinEnrichments extends React.Component<IProteinEnrichme
         this.props.store.setProteinEnrichmentProfile(m);
     }
 
+    private readonly enrichmentAnalysisGroups = remoteData({
+        await: () => [this.props.store._activeGroupsOverlapRemoved],
+        invoke: () => {
+            const groups = _.map(this.props.store._activeGroupsOverlapRemoved.result, group => {
```
Comment (alisman) [inline src/pages/groupComparison/ProteinEnrichments.tsx]: you can use the native map method on array.  if this is an object, then yeah, lodash map is helpful

File: src/pages/resultsView/enrichments/EnrichmentsUtil.tsx [hunk 10]
```diff
@@ -162,67 +163,63 @@ export function getAlterationRowData(
...
     selectedGenes: string[] | null): any[] {
 
-    return data.filter(alterationEnrichment => {
+    return data.filter(enrichmentDatum => {
         let result = false;
-        for (var i = 0; i < enrichedGroups.length; i++) {
-            const enrichedGroup = enrichedGroups[i];
-            let enrichedGroupAlteredPercentage = alterationEnrichment.groupsSet[enrichedGroup].alteredPercentage;
-            let res = _.reduce(alterationEnrichment.groupsSet, (acc, next, group) => {
+        for (var i = 0; i < expressedGroups.length; i++) {
```
Comment (alisman) [inline src/pages/resultsView/enrichments/EnrichmentsUtil.tsx]: lets just switch to forEach while we're at it

File: src/pages/resultsView/enrichments/EnrichmentsUtil.tsx [hunk 11]
```diff
@@ -454,6 +451,94 @@ export function getGroupColumns(groups: { name: string, description: string, col
...
+        },
+        filter: (d: ExpressionEnrichmentRow, filterString: string, filterStringUpper: string) =>
+            d.enrichedGroup.toUpperCase().includes(filterStringUpper),
+        sortBy: (d: ExpressionEnrichmentRow) => d.enrichedGroup,
+        download: (d: ExpressionEnrichmentRow) => d.enrichedGroup,
+        tooltip: <span>The group with the highest expression frequency</span>
+    }
+
+    //minimum 2 group are required for enrichment analysis
+    if (groups.length < 2) {
```
Comment (alisman) [inline src/pages/resultsView/enrichments/EnrichmentsUtil.tsx]: should we move this higher up to prevent unnecessary logic?

File: src/pages/resultsView/enrichments/ExpressionEnrichmentsContainer.tsx [hunk 12]
```diff
@@ -53,28 +48,19 @@ export default class ExpressionEnrichmentContainer extends React.Component<IExpr
     @observable.shallow checkedGenes: string[] = [];
     @observable clickedGeneHugo: string;
     @observable clickedGeneEntrez: number;
-    @observable.ref selectedGenes: string[]|null;
-    @observable.ref highlightedRow:ExpressionEnrichmentRow|undefined;
+    @observable.ref selectedGenes: string[] | null;
+    @observable.ref highlightedRow: ExpressionEnrichmentRow | undefined;
+    @observable _expressedGroups: string[] = this.props.groups.map(group => group.name);
```
Comment (alisman) [inline src/pages/resultsView/enrichments/ExpressionEnrichmentsContainer.tsx]: probably should be .ref too unless it's gonna be mutated anywhere

File: src/pages/resultsView/enrichments/ExpressionEnrichmentsContainer.tsx [hunk 13]
```diff
@@ -108,43 +94,81 @@ export default class ExpressionEnrichmentContainer extends React.Component<IExpr
     }
 
     private dataStore = new EnrichmentsTableDataStore(
-        ()=>{
+        () => {
```
Comment (alisman) [inline src/pages/resultsView/enrichments/ExpressionEnrichmentsContainer.tsx]: i prefer no spaces.

File: src/pages/resultsView/enrichments/ProteinEnrichmentsTab.tsx [hunk 14]
```diff
@@ -21,20 +22,42 @@ export default class ProteinEnrichmentsTab extends React.Component<IProteinEnric
...
-        if (this.props.store.proteinEnrichmentData.isPending) {
+        if (this.props.store.proteinEnrichmentData.isPending || this.enrichmentAnalysisGroups.isPending) {
             return <Loader isLoading={true} center={true} size={"big"} />;
-        } else if (this.props.store.proteinEnrichmentData.isError) {
-            return <ErrorMessage/>
+        } else if (this.props.store.proteinEnrichmentData.isError || this.enrichmentAnalysisGroups.isError) {
+            return <ErrorMessage />
         } else {
-
+            console.log(this.enrichmentAnalysisGroups.result)
```
Comment (alisman) [inline src/pages/resultsView/enrichments/ProteinEnrichmentsTab.tsx]: console.log

File: src/pages/resultsView/enrichments/ExpressionEnrichmentsContainer.tsx [hunk 15]
```diff
@@ -108,43 +94,81 @@ export default class ExpressionEnrichmentContainer extends React.Component<IExpr
...
+        return getExpressionEnrichmentColumns(this.props.groups, this.props.alteredVsUnalteredMode);
+    }
+
+    @computed get visibleOrderedColumnNames() {
+        const columns = [];
+        columns.push(ExpressionEnrichmentTableColumnType.GENE,
+            ExpressionEnrichmentTableColumnType.CYTOBAND);
+
+        this.props.groups.forEach(group => {
+            columns.push(group.name + " mean");
```
Comment (adamabeshouse) [inline src/pages/resultsView/enrichments/ExpressionEnrichmentsContainer.tsx]: mean and standard deviation I think should be wrapped in a function that generates column name, or use an enum value like `ExpressionEnrichmentTableColumnType.MEAN_SUFFIX`

File: src/pages/groupComparison/GroupComparisonStore.ts [hunk 16]
```diff
@@ -713,7 +716,7 @@ export default class GroupComparisonStore {
             (this.activeGroups.isComplete && this.activeGroups.result!.length === 0 && this.tabHasBeenShown.get(GroupComparisonTab.SURVIVAL));
     }
 
-    @computed get survivalTabGrey() {
```
Comment (adamabeshouse) [inline src/pages/groupComparison/GroupComparisonStore.ts]: why did you change it from Grey? I like Grey rather than disabled because the tab can still be clicked

File: src/pages/groupComparison/GroupComparisonStore.ts [hunk 17]
```diff
@@ -748,7 +751,7 @@ export default class GroupComparisonStore {
             (this.activeGroups.isComplete && this.activeGroups.result!.length === 0 && this.tabHasBeenShown.get(GroupComparisonTab.CNA));
     }
 
-    @computed get copyNumberTabGrey() {
+    @computed get copyNumberDisabld() {
```
Comment (adamabeshouse) [inline src/pages/groupComparison/GroupComparisonStore.ts]: Disabld

File: src/pages/groupComparison/GroupComparisonStore.ts [hunk 18]
```diff
@@ -664,18 +654,24 @@ export default class GroupComparisonStore {
     }
 
     readonly mRNAEnrichmentData = makeEnrichmentDataPromise({
-        await:()=>[this.enrichmentsGroup1, this.enrichmentsGroup2,this.mRNAEnrichmentProfile],
+        await:()=>[this.mRNAEnrichmentProfile, this._activeGroupsOverlapRemoved],
```
Comment (adamabeshouse) [inline src/pages/groupComparison/GroupComparisonStore.ts]: I think it should use activeGroups, rather than activeGroupsOverlapRemoved. I made this change to mutations and cna tabs in an upcoming PR of mine.

File: src/pages/groupComparison/GroupComparisonStore.ts [hunk 19]
```diff
@@ -684,18 +680,25 @@ export default class GroupComparisonStore {
     });
 
     readonly proteinEnrichmentData = makeEnrichmentDataPromise({
-        await:()=>[this.enrichmentsGroup1, this.enrichmentsGroup2,this.proteinEnrichmentProfile],
+        await:()=>[this.proteinEnrichmentProfile, this._activeGroupsOverlapRemoved],
```
Comment (adamabeshouse) [inline src/pages/groupComparison/GroupComparisonStore.ts]: use activeGroups instead

File: src/pages/groupComparison/MRNAEnrichments.tsx [hunk 20]
```diff
@@ -21,31 +23,37 @@ export default class MRNAEnrichments extends React.Component<IMRNAEnrichmentsPro
         this.props.store.setMRNAEnrichmentProfile(m);
     }
 
-    readonly tabUI = MakeEnrichmentsTabUI(()=>this.props.store, ()=>this.enrichmentsUI, "mRNA");
+    private readonly enrichmentAnalysisGroups = remoteData({
+        await: () => [this.props.store._activeGroupsOverlapRemoved],
```
Comment (adamabeshouse) [inline src/pages/groupComparison/MRNAEnrichments.tsx]: activeGroups

File: src/pages/groupComparison/ProteinEnrichments.tsx [hunk 21]
```diff
@@ -21,31 +23,37 @@ export default class ProteinEnrichments extends React.Component<IProteinEnrichme
         this.props.store.setProteinEnrichmentProfile(m);
     }
 
+    private readonly enrichmentAnalysisGroups = remoteData({
+        await: () => [this.props.store._activeGroupsOverlapRemoved],
```
Comment (adamabeshouse) [inline src/pages/groupComparison/ProteinEnrichments.tsx]: activeGroups

File: src/pages/groupComparison/GroupComparisonStore.ts (hunk 16 above)
Comment (kalletlak) [reply src/pages/groupComparison/GroupComparisonStore.ts]: from @alisman comment https://github.com/cBioPortal/cbioportal-frontend/pull/2559#discussion_r306450700

Comment (adamabeshouse) [reply src/pages/groupComparison/GroupComparisonStore.ts]: ah okay. I agree with @alisman 's comment. But I think disabled is the wrong word because the tab isn't really disabled. maybe survivalTabNotRendered? survivalTabUnavailable? something else? what do you think

Comment (kalletlak) [reply src/pages/groupComparison/GroupComparisonStore.ts]: `survivalTabUnavailable` seems reasonable
